File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/main/DownloadsFragment.java
package com.github.doomsdayrs.apps.shosetsu.ui.main;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.github.doomsdayrs.apps.shosetsu.R;
import com.github.doomsdayrs.apps.shosetsu.backend.Download_Manager;
import com.github.doomsdayrs.apps.shosetsu.backend.SettingsController;
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database;
import com.github.doomsdayrs.apps.shosetsu.ui.adapters.DownloadAdapter;
import com.github.doomsdayrs.apps.shosetsu.variables.DownloadItem;
import com.github.doomsdayrs.apps.shosetsu.variables.Settings;
import com.github.doomsdayrs.apps.shosetsu.variables.Statics;

import java.util.ArrayList;
import java.util.List;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 * ====================================================================
 * Shosetsu
 * 9 / June / 2019
 *
 * @author github.com/doomsdayrs
 * @author github.com/hXtreme
 */
//TODO selection mechanic with options to delete,  pause,  and more
public class DownloadsFragment extends Fragment {
    public static List<DownloadItem> downloadItems = new ArrayList<>();

    @SuppressLint("StaticFieldLeak")
    private static RecyclerView recyclerView;
    private static DownloadAdapter adapter;

    /**
     * Constructor
     */
    public DownloadsFragment() {
        setHasOptionsMenu(true);
    }


    private static void refreshList() {
        if (DownloadsFragment.adapter != null)
            if (DownloadsFragment.adapter.downloadsFragment != null)
                if (DownloadsFragment.adapter.downloadsFragment.getActivity() != null)
                    DownloadsFragment.adapter.downloadsFragment.getActivity().runOnUiThread(() -> adapter.notifyDataSetChanged());
    }

    public static void removeDownloads(DownloadItem downloadItem) {
        for (int x = 0; x < downloadItems.size(); x++)
            if (downloadItems.get(x).chapterURL.equals(downloadItem.chapterURL)) {
                downloadItems.remove(x);
                return;
            }

        refreshList();
    }

    public static void markError(DownloadItem d) {
        for (DownloadItem downloadItem : downloadItems)
            if (downloadItem.chapterURL.equals(d.chapterURL))
                d.setStatus("Error");
        refreshList();
    }

    public static void toggleProcess(DownloadItem d) {
        for (DownloadItem downloadItem : downloadItems)
            if (downloadItem.chapterURL.equals(d.chapterURL))
                if (downloadItem.getStatus().equals("Pending") || downloadItem.getStatus().equals("Error"))
                    downloadItem.setStatus("Downloading");
                else downloadItem.setStatus("Pending");
        refreshList();
    }


    @Override
    public void onResume() {
        super.onResume();
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        recyclerView = null;
        adapter = null;
    }

    /**
     * Save data of view before destroyed
     *
     * @param outState output save
     */
    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    /**
     * Creates view
     *
     * @param inflater           inflater to retrieve objects
     * @param container          container of this fragment
     * @param savedInstanceState save
     * @return View
     */
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        Log.d("OnCreateView", "NovelFragmentChapters");
        Statics.mainActionBar.setTitle("Downloads");
        View view = inflater.inflate(R.layout.fragment_downloads, container, false);
        recyclerView = view.findViewById(R.id.fragment_downloads_recycler);
        downloadItems = Database.DatabaseDownloads.getDownloadList();
        setDownloads();
        return view;
    }

    /**
     * Sets the novel chapters down
     */
    private void setDownloads() {
        recyclerView.setHasFixedSize(false);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext());
        adapter = new DownloadAdapter(this);
        adapter.setHasStableIds(true);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);
    }

    /**
     * Creates the option menu (on the top toolbar)
     *
     * @param menu     Menu reference to fill
     * @param inflater Object to inflate the menu
     */
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.toolbar_downloads, menu);
        MenuItem menuItem = menu.findItem(R.id.toolbar_downloads_pause);
        if (Settings.downloadPaused)
            menuItem.setIcon(R.drawable.ic_pause_circle_filled_black_24dp);

        menuItem.setOnMenuItemClickListener(a -> {
            if (SettingsController.INSTANCE.togglePause())
                a.setIcon(R.drawable.ic_pause_circle_filled_black_24dp);
            else {
                a.setIcon(R.drawable.ic_pause_circle_outline_black_24dp);
                Download_Manager.init();
            }
            return true;
        });

    }

}

File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/ui/downloads/DownloadsFragment.kt
package com.github.doomsdayrs.apps.shosetsu.ui.downloads

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.doomsdayrs.apps.shosetsu.R
import com.github.doomsdayrs.apps.shosetsu.backend.DownloadManager.initDownloadManager
import com.github.doomsdayrs.apps.shosetsu.backend.Utilities
import com.github.doomsdayrs.apps.shosetsu.backend.database.Database
import com.github.doomsdayrs.apps.shosetsu.ui.downloads.adapters.DownloadAdapter
import com.github.doomsdayrs.apps.shosetsu.variables.Broadcasts
import com.github.doomsdayrs.apps.shosetsu.variables.DownloadItem
import com.github.doomsdayrs.apps.shosetsu.variables.Settings
import kotlinx.android.synthetic.main.fragment_downloads.*

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 * Shosetsu
 * 9 / June / 2019
 *
 * @author github.com/doomsdayrs
 */
//TODO selection mechanic with options to delete,  pause,  and more
class DownloadsFragment : Fragment(R.layout.fragment_downloads) {
    var downloadItems: ArrayList<DownloadItem> = ArrayList()
    var adapter: DownloadAdapter = DownloadAdapter(this)
    private lateinit var receiver: BroadcastReceiver

    init {
        setHasOptionsMenu(true)
    }

    override fun onDestroy() {
        super.onDestroy()
        activity?.unregisterReceiver(receiver)
    }

    override fun onResume() {
        super.onResume()
        adapter.notifyDataSetChanged()
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Utilities.setActivityTitle(activity, getString(R.string.downloads))
        downloadItems = Database.DatabaseDownloads.getDownloadList()
        fragment_downloads_recycler.setHasFixedSize(false)
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(context)
        adapter = DownloadAdapter(this)
        adapter.setHasStableIds(true)
        fragment_downloads_recycler.layoutManager = layoutManager
        fragment_downloads_recycler.adapter = adapter
        val filter = IntentFilter()

        filter.addAction(Broadcasts.BROADCAST_NOTIFY_DATA_CHANGE)
        filter.addAction(Broadcasts.DOWNLOADS_MARK_ERROR)
        filter.addAction(Broadcasts.DOWNLOADS_TOGGLE)
        filter.addAction(Broadcasts.DOWNLOADS_REMOVE)

        receiver = object : BroadcastReceiver() {


            private fun removeDownloads(chapterURL: String) {
                for (x in adapter.downloadsFragment.downloadItems.indices) if (adapter.downloadsFragment.downloadItems[x].chapterURL == chapterURL) {
                    adapter.downloadsFragment.downloadItems.removeAt(x)
                    return
                }
                adapter.notifyDataSetChanged()
            }

            private fun markError(chapterURL: String) {
                for (downloadItem in adapter.downloadsFragment.downloadItems)
                    if (downloadItem.chapterURL == chapterURL)
                        downloadItem.status = "Error"

                fragment_downloads_recycler?.adapter?.notifyDataSetChanged()

            }

            private fun toggleProcess(chapterURL: String) {
                for (downloadItem in adapter.downloadsFragment.downloadItems)
                    if (downloadItem.chapterURL == chapterURL)
                        if (downloadItem.status == "Pending" || downloadItem.status == "Error")
                            downloadItem.status = "Downloading"
                        else downloadItem.status = "Pending"
                fragment_downloads_recycler?.adapter?.notifyDataSetChanged()
            }


            override fun onReceive(context: Context?, intent: Intent?) {
                intent?.let { i ->
                    when (i.action) {
                        Broadcasts.BROADCAST_NOTIFY_DATA_CHANGE -> (fragment_downloads_recycler?.adapter as DownloadAdapter).notifyDataSetChanged()
                        Broadcasts.DOWNLOADS_REMOVE -> i.getStringExtra(Broadcasts.DOWNLOADS_RECIEVED_URL)?.let { removeDownloads(it) }
                        Broadcasts.DOWNLOADS_TOGGLE -> i.getStringExtra(Broadcasts.DOWNLOADS_RECIEVED_URL)?.let { toggleProcess(it) }
                        Broadcasts.DOWNLOADS_MARK_ERROR -> i.getStringExtra(Broadcasts.DOWNLOADS_RECIEVED_URL)?.let { markError(it) }
                        else -> Log.e("DownloadsFragment", "No action provided!")
                    }
                }
            }

        }
        activity?.registerReceiver(receiver, filter)

    }

    /**
     * Creates the option menu (on the top toolbar)
     *
     * @param menu     Menu reference to fill
     * @param inflater Object to inflate the menu
     */
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.toolbar_downloads, menu)
        val menuItem = menu.findItem(R.id.toolbar_downloads_pause)
        if (Settings.downloadPaused)
            menuItem.setIcon(R.drawable.ic_play_circle_filled_black_24dp)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.toolbar_downloads_pause) {
            if (Utilities.togglePause())
                item.setIcon(R.drawable.ic_play_circle_filled_black_24dp)
            else {
                item.setIcon(R.drawable.ic_pause_circle_outline_black_24dp)
                initDownloadManager(activity!!)
            }
            return true
        }
        return false
    }
}